File: app/src/main/java/com/kyagamy/step/game/newplayer/Combo.java
package com.kyagamy.step.game.newplayer;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;

import com.kyagamy.step.R;
import com.kyagamy.step.common.step.CommonGame.CustomSprite.SpriteReader;
import com.kyagamy.step.common.step.CommonGame.TransformBitmap;


public class Combo {

    public  static final byte VALUE_PERFECT =0;
    public  static final byte VALUE_GREAT = 1;
    public  static final byte VALUE_GOOD = 2;
    public  static final byte VALUE_BAD = 3;
    public  static final byte VALUE_MISS = 4;
    public  static final byte VALUE_MISSING = 1;

    Long timeMark;
    short auxLife = 1;

    private SpriteReader judgeSprite, numberCombo;
    private Bitmap comboImage, badCombo, currentBitMapCombo;
    private float ratioStep, aument, ratioStepCombo, aumentLabel, tiemposcombo;
    private int posIntY,x,y, posIntXCombo, posIntYCombo, posIntYNC, posIntX, combo = 0,aumentCombo = -220;

    public short positionJudge = 0;
    float factor=1f;

    public Combo(Context c,StepsDrawer stepsDrawer) {
        this.x=stepsDrawer.sizeX;
        this.y=stepsDrawer.sizeY;
        BitmapFactory.Options myOpt2 = new BitmapFactory.Options();
        myOpt2.inSampleSize = 0;
        numberCombo = new SpriteReader(BitmapFactory.decodeResource(c.getResources(), R.drawable.play_combo_number, myOpt2),10 , 1, 1f);
        judgeSprite = new SpriteReader( BitmapFactory.decodeResource(c.getResources(), R.drawable.play_combo_judge, myOpt2), 1, 5, 1f);
        comboImage = BitmapFactory.decodeResource(c.getResources(), R.drawable.play_combo, myOpt2);
        badCombo = BitmapFactory.decodeResource(c.getResources(), R.drawable.play_combo_bad, myOpt2);
        // dibujante.setColor(Color.TRANSPARENT);


        ////////Medidades
            /*
            Perfect = (x=20%,y=6&) y max (x=30%,y=8.4&)
            Combo   =   (x=10%,y=3.1&) y max (x=15%,y=5&)
            1000    =  (x=5%,y=5&)
            */
        /////////
        ratioStep = 0.66f;
        aument = 0.0425f;
        aumentLabel = 2 * aument / 3;
        ratioStepCombo = 2 * ratioStep / 3;


        posIntY = (int) (y / 2 - (y * 0.084) / 2);


        posIntYNC = 0;

    timeMark =System.currentTimeMillis();
    }


    public void start() {
        tiemposcombo = System.currentTimeMillis();
    }

    public void show() {

        aumentCombo = 8;
        if (combo >= 0) {
            currentBitMapCombo = comboImage;
        } else {
            currentBitMapCombo = badCombo;
        }


    }

    public void setComboUpdate(short typeTap) {
        aumentCombo = 12;
        positionJudge=typeTap;
        switch (typeTap){
            case VALUE_PERFECT:
            case VALUE_GREAT:
                combo++;
                break;
            case VALUE_BAD:
            case VALUE_MISS:
                combo = (combo >0)?0: combo--;
        }
        show();

    }

    public void update(){
        if (System.nanoTime() - timeMark > 150) {
//            if (aumentCombo > 6 || aumentCombo < 0) {
//                auxLife *= -1;
//            /}
            aumentCombo -= 1;
            timeMark = System.nanoTime();
        }

        //;

    }



    public void draw(Canvas canvas) {
        posIntX = (int) (x / 2 - (x * 0.4 * (ratioStep + aument * aumentCombo)) / 2);


        posIntXCombo = (int) (x / 2 - (x * 0.13) * (ratioStep + aument * aumentCombo) / 2);


        posIntYCombo = (int) (posIntY + (y * 0.039) + (y * 0.084) * (ratioStepCombo + aumentLabel * aumentCombo) / 2);// (int) (y / 2 - (y * 0.05) / 2);


        if (aumentCombo > -12) {
            if (aumentCombo >= 0) {

                canvas.drawBitmap(judgeSprite.frames[positionJudge], null, new Rect(posIntX, posIntY, posIntX + (int) ((x * 0.4) * (ratioStep + aument * aumentCombo)), posIntY + (int) ((y * 0.084*factor) * (ratioStep + aument * aumentCombo))), new Paint());
                posIntYNC = posIntYCombo + (int) ((comboImage.getHeight()) * (ratioStepCombo + aumentLabel * aumentCombo) * 0.7);

            } else if (aumentCombo < -6) {
                posIntX = (int) (x / 2 - (x * 0.4 * (ratioStep)) / 2);
                int opacidad = (100 + 5 * aumentCombo);
                int xt = aumentCombo * -1;
                int yt = aumentCombo;
                canvas.drawBitmap(TransformBitmap.makeTransparent(judgeSprite.frames[positionJudge], opacidad), null, new Rect(posIntX, posIntY, posIntX + (int) (x * 0.4 * (ratioStep) * ((float) xt * 5 / 100)), 10 * aumentCombo + posIntY + (int) ((y *factor* 0.084) * (ratioStep + aument * aumentCombo) * ((float) yt * 5 / 100))), new Paint());

            } else {
                posIntX = (int) (x / 2 - (x * 0.4 * (ratioStep)) / 2);
                canvas.drawBitmap(judgeSprite.frames[positionJudge], null, new Rect(posIntX, posIntY, posIntX + (int) (x * 0.4 * ratioStep), posIntY + (int) (y * 0.084 *factor* ratioStep)), new Paint());
                // canvas.drawBitmap(currentBitMapCombo, null, new Rect(posIntXCombo, posIntYCombo, posIntXCombo + (int) ((x * 0.17) * (ratiostepcombo + aumentolabel)), posIntYCombo + (int) ((y * 0.053) * (ratiostepcombo + aumentolabel))), new Paint());
            }

            //draw
            //canvas.drawText(""+Combo,x/2-50,y/2,dibujante);

            posIntYNC = posIntYCombo + (int) ((y * 0.05) * (ratioStepCombo + aumentLabel * aumentCombo));
            if (combo > 3 || combo < -3) {


                canvas.drawBitmap(currentBitMapCombo, null, new Rect(posIntXCombo, posIntYCombo, posIntXCombo + (int) ((x * 0.17) * (ratioStepCombo + aumentLabel * aumentCombo)), posIntYCombo + (int) ((y * 0.053) * (ratioStepCombo + aumentLabel * aumentCombo))), new Paint());

                ////

                long lc = 100000000 + Math.abs(combo);
                String sc = lc + "";
                String sc2 = Math.abs(combo) + "";
                int nveces = 4;//dice el numero de veces que tienes

                if (sc2.length() > 3) {
                    nveces = sc2.length() + 1;
                }
                for (int w = 1; w < nveces; w++) {
                    int leng = (nveces - 1) * (int) (x * 0.05);
                    int posxnc = (int) ((0.9 * (leng / 2) + x / 2) - x * 0.05 * w * 1.0);
                    int n = Integer.parseInt(sc.charAt(sc.length() - w) + "");
                    canvas.drawBitmap(numberCombo.frames[n], null, new Rect(posxnc, posIntYNC, posxnc + (int) (x * 0.05), posIntYNC + (int) (x * 0.05)), new Paint());
                }
            }

        }

    }
}

File: app/src/main/java/com/kyagamy/step/game/newplayer/Combo.kt
package com.kyagamy.step.game.newplayer

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import com.kyagamy.step.R
import com.kyagamy.step.common.step.commonGame.customSprite.SpriteReader
import com.kyagamy.step.game.newplayer.Evaluator.Companion.BAD
import com.kyagamy.step.game.newplayer.Evaluator.Companion.GOOD
import com.kyagamy.step.game.newplayer.Evaluator.Companion.GREAT
import com.kyagamy.step.game.newplayer.Evaluator.Companion.MAX_COMBO
import com.kyagamy.step.game.newplayer.Evaluator.Companion.MISS
import com.kyagamy.step.game.newplayer.Evaluator.Companion.PERFECT
import kotlin.math.abs

class Combo(c: Context, stepsDrawer: StepsDrawer) {
    private var timeMark: Long?

    private val judgeSprite: SpriteReader
    private val numberCombo: SpriteReader
    private val comboImage: Bitmap
    private val badCombo: Bitmap
    private var currentBitMapCombo: Bitmap? = null

    private val x: Int
    private val y: Int

    private var combo = 0
    private var aumentTip = -220
    private val paint = Paint()
    private var lifeBar: LifeBar? = null


    var positionJudge: Short = 0

    init {
        timeMark = System.currentTimeMillis()
        this.x = stepsDrawer.sizeX + stepsDrawer.offsetX
        this.y = stepsDrawer.sizeY
        val myOpt2 = BitmapFactory.Options()
        myOpt2.inSampleSize = 0
        numberCombo = SpriteReader(
            BitmapFactory.decodeResource(
                c.getResources(),
                R.drawable.play_combo_number,
                myOpt2
            ), 10, 1, 1f
        )
        judgeSprite = SpriteReader(
            BitmapFactory.decodeResource(
                c.getResources(),
                R.drawable.play_combo_judge,
                myOpt2
            ), 1, 5, 1f
        )
        comboImage = BitmapFactory.decodeResource(c.getResources(), R.drawable.play_combo, myOpt2)
        badCombo = BitmapFactory.decodeResource(c.getResources(), R.drawable.play_combo_bad, myOpt2)
    }

    fun setLifeBar(lifeBar: LifeBar) {
        this.lifeBar = lifeBar
    }

    fun show() {
        aumentTip = 20
        paint.setAlpha(255)
        currentBitMapCombo = if (combo >= 0) comboImage else badCombo
    }

    fun setComboUpdate(typeTap: Short) {
        positionJudge = typeTap
        when (typeTap) {
            VALUE_PERFECT -> {
                Evaluator.PERFECT = Evaluator.PERFECT + 1
                combo = if (combo < 0) 1 else (combo + 1)
            }

            VALUE_GREAT -> {
                Evaluator.GREAT = Evaluator.GREAT + 1
                combo = if (combo < 0) 1 else (combo + 1)
            }

            VALUE_GOOD -> {
                Evaluator.GOOD = Evaluator.GOOD + 1
                if (combo < -4) combo = 0
            }

            VALUE_BAD -> {
                Evaluator.BAD = Evaluator.BAD + 1
                if (combo != 0) combo = 0
            }

            VALUE_MISS -> {
                Evaluator.MISS = Evaluator.MISS + 1
                combo = if (combo > 0) 0 else (combo - 1)
            }
        }
        lifeBar!!.updateLife(typeTap, 1)
        if (combo > Evaluator.MAX_COMBO) Evaluator.MAX_COMBO = combo
        show()
    }

    fun update() {
        //  if (System.nanoTime() - timeMark > 100) {
        aumentTip -= 1
        timeMark = System.nanoTime()
        // }
    }

    fun draw(canvas: Canvas) {
        //setSizes
        val numberSizeY = (y * COMBO_NUMBER_RATIO_X).toInt()
        val numberSizeX = (y * COMBO_NUMBER_RATIO_X).toInt()

        var comboSizeY = ((y * COMBO_TEXT_RATIO_Y)).toInt()
        var comboSizeX = ((y * COMBO_TEXT_RATIO_X)).toInt()

        var labelSizeY = ((y * COMBO_LABEL_RATIO_Y)).toInt()
        var labelSizeX = ((y * COMBO_LABEL_RATIO_X)).toInt()

        //initX For each type
        if (aumentTip > 14 && aumentTip < 21) {
            val relation: Float = 1 + (aumentTip - 15) * 0.22f * RATIO_BIGGER_LABEL
            labelSizeY = (labelSizeY * relation).toInt()
            labelSizeX = (labelSizeX * relation).toInt()
            comboSizeX = (comboSizeX * ((relation - 1) / 3 + 1)).toInt()
            comboSizeY = (comboSizeY * ((relation - 1) / 3 + 1)).toInt()
        }

        val posLabelIntX = ((x / 2f - labelSizeX / 2f) * 1.008).toInt()
        val posComboIntX = (x / 2f - comboSizeX / 2f).toInt()

        if (aumentTip < 6) paint.setAlpha(abs(-(255 / (5) * aumentTip)))


        var posIntYCombo =
            (y / 2 - (numberSizeY + labelSizeY + comboSizeY) / 2) // (int) (y / 2 - (y * 0.05) / 2);

        if (aumentTip > 0) {
            canvas.drawBitmap(
                judgeSprite.frames[positionJudge.toInt()],
                null,
                Rect(
                    posLabelIntX,
                    posIntYCombo,
                    posLabelIntX + labelSizeX,
                    posIntYCombo + labelSizeY
                ),
                paint
            )

            posIntYCombo = (posIntYCombo + labelSizeY * 1.08).toInt()
            if (combo > 3 || combo < -3) {
                //show combo text
                canvas.drawBitmap(
                    currentBitMapCombo!!,
                    null,
                    Rect(
                        posComboIntX,
                        posIntYCombo,
                        posComboIntX + comboSizeX,
                        posIntYCombo + comboSizeY
                    ),
                    paint
                )
                posIntYCombo += comboSizeY
                val stringComboAux = (100000000 + abs(combo)).toString() + ""
                val stringCombo = abs(combo).toString() + ""

                var drawTimes =
                    4 //number of types you need to draw number example combo 39 then 3 digits show 039
                if (stringCombo.length > 3) drawTimes = stringCombo.length + 1

                for (w in 1..<drawTimes) {
                    val totalComboLength = (drawTimes - 1) * numberSizeX
                    val positionCurrentNumber = ((totalComboLength / 2) + x / 2) - numberSizeX * w
                    val n = (stringComboAux.get(stringComboAux.length - w).toString() + "").toInt()
                    canvas.drawBitmap(
                        numberCombo.frames[n],
                        null,
                        Rect(
                            positionCurrentNumber,
                            posIntYCombo,
                            positionCurrentNumber + numberSizeX,
                            posIntYCombo + numberSizeY
                        ),
                        paint
                    )
                }
            }
        }
    }

    companion object {
        const val VALUE_PERFECT: Short = 0
        const val VALUE_GREAT: Short = 1
        const val VALUE_GOOD: Short = 2
        const val VALUE_BAD: Short = 3
        const val VALUE_MISS: Short = 4
        const val VALUE_MISSING: Short = 1


        //proportions Y
        private val COMBO_TEXT_RATIO_X = 0.14815f * 1.25f
        private val COMBO_TEXT_RATIO_Y = 0.0363637f * 1.25f

        private val COMBO_NUMBER_RATIO_X = 0.05555556f * 1.15f
        private const val COMBO_NUMBER_RATIO_Y = 0.06141616f

        private const val COMBO_LABEL_RATIO_X = 0.306f
        private const val COMBO_LABEL_RATIO_Y = 0.0555555556f

        private const val RATIO_BIGGER_LABEL = 0.6666666667f
    }
}
